src/parser/ParsedQuery.cpp

```diff
@@ -405,48 +412,89 @@ ParsedQuery::GraphPatternOperation::GraphPatternOperation(
                  "Union expects two sub graph patterns.");
       }
       break;
+    default:
+      AD_THROW(ad_semsearch::Exception::CHECK_FAILED,
+               "This constructor should only be used for UNION and OPTIONAL "
+               "type operations.");
   }
   _childGraphPatterns.insert(_childGraphPatterns.end(), children.begin(),
                              children.end());
 }
 
+ParsedQuery::GraphPatternOperation::GraphPatternOperation(
+    ParsedQuery::GraphPatternOperation::Type type)
+    : _type(type) {
+  switch (_type) {
+    case Type::SUBQUERY:
+      _subquery = nullptr;
+      break;
+    default:
+      AD_THROW(ad_semsearch::Exception::CHECK_FAILED,
+               "This constructor should only be used for SUBQUERY"
+               "type operations.");
+  }
+}
+
 // _____________________________________________________________________________
 ParsedQuery::GraphPatternOperation::GraphPatternOperation(
     GraphPatternOperation&& other)
-    : _type(other._type),
-      _childGraphPatterns(std::move(other._childGraphPatterns)) {
-  other._childGraphPatterns.clear();
+    : _type(other._type) {
+  if (_type == Type::SUBQUERY) {
+    _subquery = other._subquery;
+    other._subquery = nullptr;
+  } else {
+    new (&_childGraphPatterns) std::vector<GraphPattern*>();
```

Comment:
Do you really need placement `new` to initialize an object in a `union` can't one just call the constructor explicitly?

Reply:
The problem is, that only the vector or the subquery may be initialized due to the union. As far as I know, any form of calling the constructor explicitly includes then using the assignment operator, which would then operate on uninitialized memory (as _childGraphPattern does not get default constructed in the union). I haven't done anything like this before though, so I'm happy for feedback on how to handle this in a better way.

Reply:
Looking at the first example [here](https://en.cppreference.com/w/cpp/language/union) the `std::string` in the union is constructed using `operator=()` wand the `std::vector<int>` is constructed with placement new just as in your code. So maybe both works and `operator=()` doesn't care if the memory is uninitialized but maybe that's only true for `std::string`? Looking at that code we may also need to add explicit destructor calls for the `std::vector<GraphPattern*>` though.

Reply:
Looks like `S s = "str";` only works because it also constructs. The following fails:
```
#include <iostream>
#include <string>
#include <vector>
 
union S
{
    std::string str;
    std::vector<int> vec;
};

int main()
{
    S s;
    s.str = {"Hello, world"};
    // at this point, reading from s.vec is undefined behavior
    std::cout << "s.str = " << s.str << '\n';
    s.str.~basic_string();
    new (&s.vec) std::vector<int>;
    // now, s.vec is the active member of the union
    s.vec.push_back(10);
    std::cout << s.vec.size() << '\n';
    s.vec.~vector();
}
```

Reply:
I added the missing destructor calls for the vector